1.1. Теоретический материал - Библиотека NumPy
NumPy (NumericalPython) - это библиотека Python с открытым 
исходным кодом, которая используется практически во всех областях науки 
и техники. Это универсальный стандарт для работы с числовыми данными в 
Python.
Если у вас уже есть Python, вы можете установить NumPy с помощью 
командной строки:
 pip install numpy
Чтобы начать использовать NumPy необходимо импортировать 
соответствующую библиотеку:
 import numpy as np
Основным объектом NumPy является однородный многомерный 
массив (в numpy называется numpy.ndarray). Это многомерный массив 
элементов (обычно чисел), одного типа.
Наиболее важные атрибуты объектов ndarray:
ndarray.ndim - число измерений (чаще их называют "оси") массива.
ndarray.shape - размеры массива, его форма. Это кортеж натуральных 
чисел, показывающий длину массива по каждой оси. Для матрицы из n строк 
и m столбов, shape будет (n,m). Число элементов кортежа shape равно ndim.
ndarray.size - количество элементов массива. Очевидно, равно 
произведению всех элементов атрибута shape.
ndarray.dtype - объект, описывающий тип элементов массива. Можно 
определить dtype, используя стандартные типы данных Python. NumPy здесь 
предоставляет целый букет возможностей, как встроенных, например: bool_, 
character, int8, int16, int32, int64, float8, float16, float32, float64, complex64, 
object_, так и возможность определить собственные типы данных, в том 
числе и составные.
ndarray.itemsize - размер каждого элемента массива в байтах.
ndarray.data - буфер, содержащий фактические элементы массива. 
Обычно не нужно использовать этот атрибут, так как обращаться к 
элементам массива проще всего с помощью индексов.
Подробнее о массивах в NumPy можно найти в официальной 
документации https://numpy.org/doc/stable/user/absolute_beginners.html

1.2.1 Пример
Задача: Создать массив 5x2. Создать массив 5x2. Вывести все значения массива, 
значение элемента с индексом (3,1) и второй столбец. Индексация 
начинается с нуля.


In [2]:
import numpy as np
x = np.array([[1, 2], [3, 4], [5, 6], [7, 8], [9, 10]])
print(x)
print(x[3][1])
print(x[1])

[[ 1  2]
 [ 3  4]
 [ 5  6]
 [ 7  8]
 [ 9 10]]
8
[3 4]


1.2.2 Пример
Задача: Пример. Выполнить следующее:
1. Создать вектор (одномерный массив) размера 10, заполненный 
нулями.
2. Создать вектор размера 10, заполненный единицами.
3. Создать вектор размера 10, заполненный заданным числом.
4. Создать вектор со значениями от 10 до 19.

In [3]:
import numpy as np
a = np.zeros(10)
b = np.ones(10)
c = np.full(10,5)
d = np.arange(10,20)
print(a, '\n', b, '\n', c, '\n', d)

[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.] 
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.] 
 [5 5 5 5 5 5 5 5 5 5] 
 [10 11 12 13 14 15 16 17 18 19]


1.2.3 Пример
Задача:Создать массив 10x10 со случайными значениями, найти минимум, 
максимум и среднее значение.

In [5]:
Z = np.random.random((10, 10))
Zmin, Zmax, Zmean = Z.min(), Z.max(), Z.mean()
print(Zmin, Zmax, Zmean)

0.017078900712722556 0.9978137032786439 0.5193149527265929


1.2.4 Пример
Задача: Задать матрицу размерности 5 на 5 и поменять 2 строки в матрице
местами.

In [6]:
A = np.arange(25).reshape(5, 5)
A[[0,1]] = A[[1,0]]
print(A)

[[ 5  6  7  8  9]
 [ 0  1  2  3  4]
 [10 11 12 13 14]
 [15 16 17 18 19]
 [20 21 22 23 24]]


1.2.5 Пример
Задача: Выяснить результат следующих выражений:
0 * np.nan
np.nan == np.nan
np.inf > np.nan
np.nan - np.nan
0.3 == 3 * 0.1

In [7]:
print(0 * np.nan)
print(np.nan == np.nan)
print(np.inf > np.nan)
print(np.nan - np.nan)
print(0.3 == 3 * 0.1)

nan
False
False
nan
False


1.2.6 Пример
Задача: Отсортировать массив

In [8]:
arr = np.array([2, 1, 5, 3, 7, 4, 6, 8])
print(np.sort(arr))

[1 2 3 4 5 6 7 8]


1.3.1 Задание
Задача: Создать 8x8 матрицу и заполнить её в шахматном порядке нулями и 
единицами.

In [15]:
import numpy as np
my_array = np.zeros(64).reshape(8,8)
for i in range (0, 8):
    for j in range (0, 8):
        if (j + i) % 2 == 1:
            my_array[i][j] = 1
print(my_array)

[[0. 1. 0. 1. 0. 1. 0. 1.]
 [1. 0. 1. 0. 1. 0. 1. 0.]
 [0. 1. 0. 1. 0. 1. 0. 1.]
 [1. 0. 1. 0. 1. 0. 1. 0.]
 [0. 1. 0. 1. 0. 1. 0. 1.]
 [1. 0. 1. 0. 1. 0. 1. 0.]
 [0. 1. 0. 1. 0. 1. 0. 1.]
 [1. 0. 1. 0. 1. 0. 1. 0.]]


1.3.2 Задание
Задача: Создать 5x5 матрицу со значениями в строках от 0 до 4. Для создания 
необходимо использовать функцию arrange.


In [31]:
import numpy as np
my_array = np.full((5, 5), np.arange(0, 5))
print(array)

[[0 1 2 3 4]
 [0 1 2 3 4]
 [0 1 2 3 4]
 [0 1 2 3 4]
 [0 1 2 3 4]]


1.3.3 Задание
Задача: Создать массив 3x3x3 со случайными значениями.

In [40]:
import numpy as np
my_array = np.random.random((3,3,3))
print(my_array)

[[[0.41566172 0.9546949  0.22735281]
  [0.26056113 0.22438763 0.46695455]
  [0.68903783 0.5438537  0.32204107]]

 [[0.24880911 0.53992934 0.02997406]
  [0.13410931 0.95402541 0.48136294]
  [0.39463435 0.67826856 0.31117394]]

 [[0.79033282 0.31233038 0.4515745 ]
  [0.08907015 0.25023526 0.90319794]
  [0.67669058 0.36532888 0.14248537]]]


1.3.4 Задание
Задача: Создать матрицу с 0 внутри, и 1 на границах.

In [ ]:
import numpy as np
print('Enter the size of matrix: ')
size = int(input())
my_array = np.full((size, size), 0)
for i in range (0, size):
    for j in range (0, size):
        if i == 0 or j == 0 or i == size - 1 or j == size - 1:
            my_array[i][j] = 1
print(my_array)


Enter the size of matrix: 


 5


[[1 1 1 1 1]
 [1 0 0 0 1]
 [1 0 0 0 1]
 [1 0 0 0 1]
 [1 1 1 1 1]]


1.3.5 Задание
Задача: Создайте массив и отсортируйте его по убыванию.

In [59]:
def res_sort(*array):
    array = sorted(array)
    array.reverse()
    return array

x = [6, 3, 7, 43, 321, 4]
print(res_sort(*x))


[321, 43, 7, 6, 4, 3]


1.3.6 Задание
Задача: Создайте матрицу, выведите ее форму, размер и размерность.

In [88]:
import numpy as np
x = np.array([[23, 12, 7, 98, 100], [1, 1, 0, 0, 0],
            [45, 102, 67, 0, 1], [21, 4, 8, 678, 99]]).reshape(4, 5)
print(x)
print('Shape: ', x.shape)
print('Size: ', x.size)
print('Ndim: ', x.ndim)

[[ 23  12   7  98 100]
 [  1   1   0   0   0]
 [ 45 102  67   0   1]
 [ 21   4   8 678  99]]
Shape:  (4, 5)
Size:  20
Ndim:  2


2.1. Теоретический материал - Библиотека Pandas
Первым шагом в любом начинании в области машинного обучения 
является введение исходных данных в систему. Исходные данные могут 
вводиться вручную, содержаться в файле или храниться в интернете в какомлибо формате. Кроме того, часто требуется получить данные из нескольких 
источников.
Библиотека pandas – это удобный и быстрый инструмент для работы 
с данными, обладающий большим функционалом. Если очень кратко, то
pandas – это библиотека, которая предоставляет очень удобные с точки 
зрения использования инструменты для хранения данных и работе с ними. 
Библиотека pandas присутствует в стандартной поставке Anaconda. 
Если же ее там нет, то его можно установить отдельно. Для этого введите 
командной строке:
 pip install pandas
Для импорта библиотеки используйте команду:
import pandas as pd
Библиотека pandas предоставляет две ключевые структуры данных:
Series и DataFrame.
Series – это одномерная структура данных, ее можно представить, как 
таблицу с одной строкой. С Series можно работать как с обычным массивом 
(обращаться по номеру индекса), и как с ассоциированным массивом, когда 
можно использовать ключ для доступа к элементам данных.
DataFrame – это двумерная структура. Идейно она очень похожа на 
обычную таблицу, что выражается в способе ее создания и работе с ее 
элементами.

2.2.1 Пример
Задача: Создать Series из списка Python, словаря Python, и массива Numpy
(установить буквенные метки для последнего).

In [10]:
import pandas as pd
lst = [1, 2, 3, 4,5]
d = {'a':1, 'b':2, 'c':3}
ndarr = np.array([1,2,3,4,5])

s1 = pd.Series(lst)
s2 = pd.Series(d)
s3 = pd.Series(ndarr, ['a', 'b', 'c', 'd', 'e'])

print(s1)
print(s2)
print(s3)

0    1
1    2
2    3
3    4
4    5
dtype: int64
a    1
b    2
c    3
dtype: int64
a    1
b    2
c    3
d    4
e    5
dtype: int64


2.2.2 Пример
Задача: Дано два Series. Напечатать их первые элементы и все элементы после 
третьего (во втором фрейме).

In [11]:
s1 = pd.Series([1,2,3,4,5], ['a', 'b', 'c', 'd', 'e'])
s2 = pd.Series([5,4,3,2,1])
print(s1['a'])
print(s2[0])
print(s2[3:])

1
5
3    2
4    1
dtype: int64


2.2.3 Пример
Задача: Создайте новый фрейм данных.

In [12]:
df = pd.DataFrame()
df['Имя'] = ['Джеки Дженсон', 'Стивен Стивенсон']
df['Возраст'] = [38, 25]
df['Водитель'] = [True, False]
df

Имя  Возраст  Водитель
0     Джеки Дженсон       38      True
1  Стивен Стивенсон       25     False

2.2.4 Пример
Задача: Загрузите фрейм данных по ссылке: 
https://raw.githubusercontent.com/chrisalbon/simulated_datasets/master/titan
ic.csv

In [15]:
url = 'https://raw.githubusercontent.com/chrisalbon/simulated_datasets/master/titanic.csv'
df = pd.read_csv(url)
df.head(5)

Name PClass    Age     Sex  \
0                   Allen, Miss Elisabeth Walton    1st  29.00  female   
1                    Allison, Miss Helen Loraine    1st   2.00  female   
2            Allison, Mr Hudson Joshua Creighton    1st  30.00    male   
3  Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.00  female   
4                  Allison, Master Hudson Trevor    1st   0.92    male   

   Survived  SexCode  
0         1        1  
1         0        1  
2         0        0  
3         0        1  
4         1        0

2.2.5 Пример
Задача: Пронализировать характеристики фрейма данных.

In [16]:
url = 'https://raw.githubusercontent.com/chrisalbon/simulated_datasets/master/titanic.csv'
df = pd.read_csv(url)
df.head(2)
df.tail(3)
df.shape
df.describe()

Age     Survived      SexCode
count  756.000000  1313.000000  1313.000000
mean    30.397989     0.342727     0.351866
std     14.259049     0.474802     0.477734
min      0.170000     0.000000     0.000000
25%     21.000000     0.000000     0.000000
50%     28.000000     0.000000     0.000000
75%     39.000000     1.000000     1.000000
max     71.000000     1.000000     1.000000

2.2.6 Пример
Задача: Выберите индивидуальные данные или срезы фрейма данных

In [17]:
url = 'https://raw.githubusercontent.com/chrisalbon/simulated_datasets/master/titanic.csv'
df = pd.read_csv(url)
df.iloc[1:4]

Name PClass   Age     Sex  \
1                    Allison, Miss Helen Loraine    1st   2.0  female   
2            Allison, Mr Hudson Joshua Creighton    1st  30.0    male   
3  Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.0  female   

   Survived  SexCode  
1         0        1  
2         0        0  
3         0        1

2.2.7 Пример
Требуется отобрать строки фрейма данных на основе некоторого 
условия. Необходимо сформировать новый фрейм данных из пассажиров 
первого класса.

In [19]:
df[df['PClass'] == '1st'].head(2)

Name PClass   Age     Sex  Survived  SexCode
0  Allen, Miss Elisabeth Walton    1st  29.0  female         1        1
1   Allison, Miss Helen Loraine    1st   2.0  female         0        1

2.3.1 Задание
Задача: Найдитн евклидово расстояние между двумя Series (точками) a и b, не 
используя встроенную формулу.

In [89]:
import pandas as pd
a = [5, 3]  #[x_coord, y_coord]
b = [-3, 4]
s1 = pd.Series(a)
s2 = pd.Series(b)
distance = ((s2[0] - s1[0]) ** 2 + (s2[1] - s1[0]) ** 2) ** 0.5
print(distance)

8.06225774829855


2.3.2 Задание
Задача: Найдите в Интернете ссылку на любой csv файл и сформируйте из него 
фрейм данных (например, коллекцию фреймов данных можно найти 
здесь: https://github.com/akmand/datasets).

In [100]:
import pandas as pd
url = 'https://raw.githubusercontent.com/chrisalbon/simulated_datasets/master/titanic.csv'
dataframe = pd.read_csv(url)
dataframe


Name PClass    Age     Sex  \
0                      Allen, Miss Elisabeth Walton    1st  29.00  female   
1                       Allison, Miss Helen Loraine    1st   2.00  female   
2               Allison, Mr Hudson Joshua Creighton    1st  30.00    male   
3     Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.00  female   
4                     Allison, Master Hudson Trevor    1st   0.92    male   
...                                             ...    ...    ...     ...   
1308                             Zakarian, Mr Artun    3rd  27.00    male   
1309                         Zakarian, Mr Maprieder    3rd  26.00    male   
1310                               Zenni, Mr Philip    3rd  22.00    male   
1311                               Lievens, Mr Rene    3rd  24.00    male   
1312                                 Zimmerman, Leo    3rd  29.00    male   

      Survived  SexCode  
0            1        1  
1            0        1  
2            0        0  
3            0        1  
4            1        0  
...        ...      ...  
1308         0        0  
1309         0        0  
1310         0        0  
1311         0        0  
1312         0        0  

[1313 rows x 6 columns]

2.3.3 Задание
Задача: Проделайте с получившемся из предыдущего задания фреймом данных 
те же действия, что и в примерах 2.2.5-2.2.7

In [134]:
import pandas as pd
from sklearn import preprocessing
dataframe = pd.read_csv('https://raw.githubusercontent.com/akmand/datasets/master/iris.csv')
dataframe
#scaled_feature


sepal_length_cm  sepal_width_cm  petal_length_cm  petal_width_cm  \
0                5.1             3.5              1.4             0.2   
1                4.9             3.0              1.4             0.2   
2                4.7             3.2              1.3             0.2   
3                4.6             3.1              1.5             0.2   
4                5.0             3.6              1.4             0.2   
..               ...             ...              ...             ...   
145              6.7             3.0              5.2             2.3   
146              6.3             2.5              5.0             1.9   
147              6.5             3.0              5.2             2.0   
148              6.2             3.4              5.4             2.3   
149              5.9             3.0              5.1             1.8   

       species  
0       setosa  
1       setosa  
2       setosa  
3       setosa  
4       setosa  
..         ...  
145  virginica  
146  virginica  
147  virginica  
148  virginica  
149  virginica  

[150 rows x 5 columns]

3.1 Теоретический материал - работа с числовыми данными
Количественные данные что-то измеряют – будь то размер класса, 
ежемесячные продажи или оценки учащихся. Естественным способом 
представления этих величин является численным (например, 150 студентов, 
$529 392 продаж).
Нормализация данных — это общепринятая задача предобработки в 
машинном обучении. Многие алгоритмы предполагают, что все признаки 
находятся в единой шкале, как правило, от 0 до 1 или от -1 до 1. 
Существует множество способов нормализации значений признаков, 
чтобы масштабировать их к единому диапазону и использовать в различных 
моделях машинного обучения. В зависимости от используемой функции, их 
можно разделить на 2 большие группы: линейные и нелинейные. При 
нелинейной нормализации в расчетных соотношениях используются 
функции логистической сигмоиды или гиперболического тангенса. В 
линейной нормализации изменение переменных осуществляется 
пропорционально, по линейному закону.
На практике наиболее распространены следующие методы 
нормализации признаков:
 Минимакс – линейное преобразование данных в диапазоне [0..1], где 
минимальное и максимальное масштабируемые значения 
соответствуют 0 и 1 соответственно;
 Z-масштабирование данных на основе среднего значения и 
стандартного отклонения: производят деление разницы между 
переменной и средним значением на стандартное отклонение.
При масштабировании данных мы будем использовать одну из 
популярных библиотек машинного обучения Scikit-learn. Библиотека 
содержит пакет sklearn.preprocessing, который предоставляет широкие
возможности для нормализации данных. Следует отметить, что в целом 
алгоритмы обучения выигрывают от стандартизации набора данных.

Пример 3.2.1
Задача: Прошкалируйте числовой признак в диапазон между двумя значениями.

In [20]:
import numpy as np
from sklearn import preprocessing
feature = np.array([[-500.5], [-100.1], [0], [100.1], [900.9]])
minmax_scale = preprocessing.MinMaxScaler(feature_range = (0, 1))
scaled_feature = minmax_scale.fit_transform(feature)
scaled_feature

array([[0.        ],
       [0.28571429],
       [0.35714286],
       [0.42857143],
       [1.        ]])

3.2.2 Пример
Задача: Преобразуйте признак, чтобы он имел среднее значение 0 и стандартное 
отклонение 1.

In [22]:
x = np.array([[-1000.1], [-200.2], [500.5], [600.6], [9000.9]])
scaler = preprocessing.StandardScaler()
stand = scaler.fit_transform(x)
stand

array([[-0.76058269],
       [-0.54177196],
       [-0.35009716],
       [-0.32271504],
       [ 1.97516685]])

3.2.2 Пример
Задача: Дан фрейм данных 
dfTest = pd.DataFrame({'A':[14.00,90.20,90.95,96.27,91.21],
 'B':[103.02,107.26,110.35,114.23,114.68],
 'C':['big','small','big','small','small']})
Необходимо масштабировать его числовые столбцы.


In [23]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
dfTest = pd.DataFrame({'A':[14.00,90.20,90.95,96.27,91.21],
    'B':[103.02,107.26,110.35,114.23,114.68],
    'C':['big','small','big','small','small']})

dfTest[['A', 'B']] = scaler.fit_transform(dfTest[['A', 'B']])
dfTest

A         B      C
0  0.000000  0.000000    big
1  0.926219  0.363636  small
2  0.935335  0.628645    big
3  1.000000  0.961407  small
4  0.938495  1.000000  small

3.3.2 Задание
Задача: Загрузить фрейм данных по ссылке:
https://raw.githubusercontent.com/akmand/datasets/master/iris.csv.
Необходимо выполнить нормализацию первого числового признака 
(sepal_length_cm) с использованием минимаксного преобразования, а 
второго (sepal_width_cm) с задействованием z-масштабирования.

In [171]:
import pandas as pd
import numpy as np
from sklearn import preprocessing
df_1 = pd.read_csv('https://raw.githubusercontent.com/akmand/datasets/master/iris.csv', usecols = ['sepal_length_cm'])
df_2 = pd.read_csv('https://raw.githubusercontent.com/akmand/datasets/master/iris.csv', usecols = ['sepal_width_cm'])
minmax_scale = preprocessing.MinMaxScaler(feature_range = (0, 1))
z_scaler = preprocessing.StandardScaler()
names_1 = df_1.columns
names_2 = df_2.columns
scaled_feature_1 = minmax_scale.fit_transform(df_1)
scaled_feature_2 = z_scaler.fit_transform(df_2)
scaled_df_1 = pd.DataFrame(scaled_feature_1, columns = names_1)
scaled_df_2 = pd.DataFrame(scaled_feature_2, columns = names_2)
print(scaled_df_1)
print(scaled_df_2)
print('Average: ', round(scaled_df_2.mean()))
print('Standard deviation', scaled_df_2.std())



     sepal_length_cm
0           0.222222
1           0.166667
2           0.111111
3           0.083333
4           0.194444
..               ...
145         0.666667
146         0.555556
147         0.611111
148         0.527778
149         0.444444

[150 rows x 1 columns]
     sepal_width_cm
0          1.032057
1         -0.124958
2          0.337848
3          0.106445
4          1.263460
..              ...
145       -0.124958
146       -1.281972
147       -0.124958
148        0.800654
149       -0.124958

[150 rows x 1 columns]
Average:  sepal_width_cm   -0.0
dtype: float64
Standard deviation sepal_width_cm    1.00335
dtype: float64
